crates/bevy_utils/src/label.rs

```diff
@@ -47,6 +50,39 @@ where
     }
 }
 
+/// Trait for implementors of `*Label` types that support downcasting.
+pub trait LabelDowncast {
+    // FIXME: use "return position impl Trait in traits" when that stabilizes.
+    /// The type returned from [`downcast_from`](#method.downcast_from).
+    type Output: Deref<Target = Self>;
+    /// Attempts to convert data from a label to type `Self`. Returns a reference-like type.
+    fn downcast_from(data: u64) -> Option<Self::Output>;
+}
+
+#[doc(hidden)]
+pub struct VTable {
+    // FIXME: When const TypeId stabilizes, inline the type instead of using a fn pointer for indirection.
+    pub ty: fn() -> ::std::any::TypeId,
+    pub fmt: fn(u64, &mut ::std::fmt::Formatter) -> ::std::fmt::Result,
+}
+
+impl PartialEq for VTable {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        (self.ty)() == (other.ty)()
+    }
+}
+impl Eq for VTable {}
+
+impl Hash for VTable {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (self.ty)().hash(state);
+    }
+}
+
+#[doc(hidden)]
+pub static STR_INTERN: Interner<&str> = Interner::new();
```

Comment:
This needs a comment justifying the use of a persistent static IMO.

Reply:
Do you mean static interning in general, or for strings specifically?

Reply:
Static interning in general. It's a bit of an exotic pattern :)

Reply:
Do you think the module-level docs I added are sufficient justification?

Reply:
Yep, I'm thrilled with those.